x-pack/plugins/visualization_lens/public/types.ts

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+export interface EditorFrameAPI {
+  render: (domElement: Element) => void;
+  registerDatasource: (datasource: Datasource<unknown>) => void;
+  registerVisualization: (visualization: Visualization<unknown>) => void;
+}
+
+export interface EditorFrameState {
+  visualizations: { [id: string]: object };
+  datasources: { [id: string]: object };
+
+  activeDatasourceId: string;
+  activeVisualizationId: string;
+}
+
+// Hints the default nesting to the data source. 0 is the highest priority
+export type DimensionPriority = 0 | 1 | 2;
+
+// For switching between visualizations and correctly matching columns
+export type DimensionRole =
+  | 'splitChart'
+  | 'series'
+  | 'primary'
+  | 'secondary'
```

Comment:
Should we have a tertiary one?

Reply:
If we have 3D charts, they can use a string type indicating their tertiary dimension. Otherwise, we're expecting that all 2D visualizations have primary/secondary which are separate from the splitChart and series dimensions. Duplicates would be allowed.